keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1377 @@
+determine the most eligible leader out of a set of LeaseCandidate objects. This
+is a feature that could be extended to other use cases if a specific selection
+strategy is required. By default, CLE will default to the lowest compatibility
+version to respect [Kubernetes component version skew
+policy](https://kubernetes.io/releases/version-skew-policy/).
+
+Strategy is an option specified by a `LeaseCandidate` object, in most cases the
+controller responsible renewing the `LeaseCandidate` lease. The value for the
+strategy between different `LeaseCandidate` objects leading the same `Lease`
+should be the same, but during mixed version states, there is a possibility that
+they may differ. We will use a consensus protocol that favors the algorithm with
+the highest priority. The priority is a fixed list that is predetermined. For
+now, this is `NoCoordination` > `MinimumCompatibilityVersion`. For example, if
+three `LeaseCandidate` objects exist and two objects select
+`MinimumCompatibilityVersion` while the third selects `NoCoordination`,
+`NoCoordination` will take precedent and the coordinated leader election
+controller will use `NoCoordination` as the election strategy. The final
+strategy used will be written to the `Lease` object when the CLE controller
+creates the `Lease` for a suitable leader. This has the benefit of providing
+better debugging information and allows short circuiting of an election if the
+set of candidates and selected strategy is the same as before.
+
+#### Alternative for Strategy
+
+The current selected approach for handling `Strategy` has the complexity of components not having consensus and the CLE controller having to tiebreak decisions. This section highlights some alternatives
+
+Instead of having the LeaseCandidate controllers select the strategy, have the CLE controller directly select the strategy. This has the benefit that no client-go change is needed and bug fixes will land quicker via only kube-apiserver changes. Because the Lease object is created by the CLE controller, it must retrieve the default value for `Strategy` from somewhere. There are a couple other ways to populate the default value instead of via  the `LeaseCandidate`.
+
+1. Via writes to the `Strategy` field in Lease
+
+If the CLE controller always defaults to a value, a user may update the Lease
+object to take on a different strategy.
+
+```
+kind: Lease
+spec:
+  strategy: "MaximumCompatibilityVersion"
+```
+
+For instance, the spec above may be server side applied after the creation of a
+lease to change the strategy. The main drawback of this approach is that the
+apply operation may only be performed AFTER the creation of a lease. Thus, there
+will be some time that passes between when the strategy is updated from the
+default strategy. There are concerns for having an ephemeral leader until the
+strategy is fixed by the apply. This would involve the least amount of API
+changes though.
```

Comment:
Can't one create a leader-less lease?

Reply:
That's an interesting point. Essentially the candidates will make some kind of signal (assuming via annotation) that the CLE controller should create a Lease but not assign a leader.

I'm having some trouble envisioning concretely what that would look like in practice though. We'd need some kind of guarantee that the Lease will transition from leaderless to having a leader in a certain amount of time. If we rely on the user to update the strategy, that seems prone to delays and prolonged period of time with no leader. If it's done automatically, then it seems equivalent to one of the other methods where some component would be responsible for relaying the strategy to the Lease object. 

Do we have concrete use cases that we can walk through for users that don't want the default strategy of min compatibility version? Maybe that can guide us on the guarantee we provide to users.

Reply:
I was also wondering why the lease resource couldn't be created with the strategy set but in a not-yet-in-use-state.  If it's clearly expired when created, I don't see any problem creating it early for the purposes of setting a strategy.

Note that even if we did decide to go with a design where a separate resource is used to declare what strategy to sue for a lease, we still have ordering problems, and I'm not convinced the ordering problems we'd have if we introduce a new resource are any better than the ordering problems we have if we just put the strategy directly in the lease..  I currently favor this approach.

Reply:
Updated with discussion with Joe. Now favoring only having the strategy in the `Lease` object.

@sttts: One difference from what we originally discussed is that we're favoring setting the strategy via a third party rather than having the lease candidates set them. This will completely avoid the ordering problem.